pkg/cloud/services/eks/addons.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	ekscontrolplanev1 "sigs.k8s.io/cluster-api-provider-aws/controlplane/eks/api/v1alpha3"
+	eksaddons "sigs.k8s.io/cluster-api-provider-aws/pkg/eks/addons"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/record"
+)
+
+func (s *Service) reconcileAddons(ctx context.Context) error {
+	s.scope.Info("Reconciling EKS addons")
+
+	eksClusterName := s.scope.KubernetesClusterName()
+
+	// 1. get addons from AWS for cluster
+	s.scope.V(2).Info("getting installed eks addons", "cluster", eksClusterName)
+	installed, err := s.getClusterAddons()
+	if err != nil {
+		return fmt.Errorf("getting installed eks addons: %w", err)
+	}
+
+	// 2. get addons from spec
+	desiredAddons := s.translateToSdkAddon(s.scope.Addons())
+
+	// 3. Exit early if there is none
+	if len(installed) == 0 && len(desiredAddons) == 0 {
+		s.scope.Info("no addons installed and no addons to install, no action needed")
+		return nil
+	}
+
+	// 3. compute CUD operations between 1 & 2
+	s.scope.V(2).Info("creating eks addons plan", "cluster", eksClusterName, "numdesired", len(desiredAddons), "numinstalled", len(installed))
+	addonsPlan := eksaddons.NewAddonsPlan(eksClusterName, desiredAddons, installed, s.EKSClient)
+	procedures, err := addonsPlan.Create(ctx)
+	if err != nil {
+		s.scope.Error(err, "failed creating eks addons plane")
+		return fmt.Errorf("creating eks addons plan: %w", err)
+	}
+	s.scope.V(2).Info("computed EKS addons plan", "numprocs", len(procedures))
+	if len(procedures) == 0 {
+		s.scope.Info("addon plan has no actions required, no action needed")
+		return nil
+	}
+
+	// 4. Perform required operations
+	for _, procedure := range procedures {
+		s.scope.V(2).Info("Executing addon procedure", "name", procedure.Name())
+		if err := procedure.Do(ctx); err != nil {
+			s.scope.Error(err, "failed executing addon procedure", "name", procedure.Name())
```

Comment:
could the addon status be set to failed in this case?

Reply:
I've added the status to the state of the addon and this will be set via the state returned by the API. So I don't want to manually set a status.  

In the scenario where there are multiple addons and one fails do you think we should carry on with the others or stop processing? At the moment we stop processing but I'm thinking we should continue on after reporting the error.